bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairControllerHandler.java

```diff
@@ -0,0 +1,1186 @@
+                    getSchedule(i);
+                }
+
+                getLightGroups();
+
+                getValves();
+
+                getSWVersion();
+            }
+        };
+
+        scheduler.execute(queryinfo);
+    }
+
+    @Override
+    public void goOffline(ThingStatusDetail detail) {
+        super.goOffline(detail);
+
+        if (syncTimeJob != null) {
+            syncTimeJob.cancel(true);
+        }
+
+        onlineController = null;
+    }
+
+    @Nullable
+    public PentairControllerCircuit getCircuitByGroupID(String groupID) {
+        for (ExpiringCache<PentairControllerCircuit> circuitItem : circuitsCache) {
+            PentairControllerCircuit circuit = circuitItem.getLastKnownValue();
+            if (circuit == null) {
+                break;
+            }
+
+            if (circuit.getGroupID().equals(groupID)) {
+                return circuit;
+            }
+        }
+
+        return null;
+    }
+
+    public int getScheduleNumber(String name) {
+        int i;
+
+        for (i = 1; i <= NUM_SCHEDULES; i++) {
+            String str = String.format(CONTROLLER_SCHEDULE, i);
+            if (str.equals(name)) {
+                return i;
+            }
+        }
+
+        return 0;
+    }
+
+    public Unit<Temperature> getUOM() {
+        return (this.uom) ? SIUnits.CELSIUS : ImperialUnits.FAHRENHEIT;
+    }
+
+    public QuantityType<Temperature> getWaterTemp() {
+        PentairControllerStatus status = this.controllerStatusCache.getLastKnownValue();
+        Objects.requireNonNull(status);
+        return new QuantityType<Temperature>(status.poolTemp, getUOM());
+    }
+
+    public @Nullable PentairControllerSchedule getScheduleByGroupID(String groupid) {
+        int scheduleNumber = getScheduleNumber(groupid);
+        if (scheduleNumber == 0) {
+            return null;
+        }
+
+        PentairControllerSchedule schedule = schedulesCache[scheduleNumber - 1].getLastKnownValue();
+
+        return schedule;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String groupId = channelUID.getGroupId();
+
+        if (groupId == null) {
+            return;
+        }
+
+        if (command instanceof RefreshType) {
+            logger.debug("handleCommand (refresh): {}", channelUID.getId());
+
+            switch (channelUID.getIdWithoutGroup()) {
+                case CONTROLLER_CIRCUITSWITCH:
+                case CONTROLLER_TEMPERATURE:
+                case CONTROLLER_AIRTEMPERATURE:
+                case CONTROLLER_SOLARTEMPERATURE:
+                case CONTROLLER_UOM:
+                case CONTROLLER_SERVICEMODE:
+                case CONTROLLER_SOLARON:
+                case CONTROLLER_HEATERON:
+                case CONTROLLER_HEATERDELAY: {
+                    PentairControllerStatus status = controllerStatusCache.getValue(() -> {
+                        getControllerStatus();
+                    });
```

Comment:
```suggestion
                    PentairControllerStatus status = controllerStatusCache.getValue(this::getControllerStatus);
```
But please change the name of the called method as per my earlier comment.